### ISSUE TITLE
[Feature] RuleForEach Property builder

### ISSUE THREAD

**Issue by jzabroski** [JeremySkinner/FluentValidation#1123]
I think the following might be much more pleasant to write:
💛 
```csharp
RuleForEach(s => s.Lines)
                .Property(s => s.DollarAmount)
                .GreaterThanOrEqualTo(0);
```

❌ 
```csharp
RuleForEach(s => s.Lines)
                .Must(s => s.DollarAmount > 0)
                .WithMessage((s, x) => $"LineNumber {s.LineNumber} {nameof(s.Lines)} must have {x.DollarAmount} greater than 0.");
```

Granted, it's not 1:1 and the bottom example has some additional context, but you get the idea of basically letting users "recurse down an object model" and re-use the already unit tested FluentValidation validation rules.

By the way, Jeremy, the overloads on WithMessage is a nice touch!  It's pretty amazing to use a library and wonder "gee, I wonder if I just type it this way if it will just work", and IT DOES!

**Comment by jzabroski** [comment]
I also just noticed that in typing the Must syntax, I have a bug where I put `>` instead of `>=`, which I think just speaks to how much I would prefer using intellisense and FluentValidation builder API

**Comment by JeremySkinner** [comment]
The usual approach with FluentValidation isn't to decare rules for nested properties inline, but to use [child validators](https://fluentvalidation.net/start#complex-properties), so this probably isn't something we'll include in the core. That being said, you can add this behaviour with an extension method if you'd like to have it in your project:

```csharp
public static IRuleBuilderInitial<TProperty, TChildProperty> Property<T, TProperty, TChildProperty>(this IRuleBuilder<T, TProperty> rule, Expression<Func<TProperty, TChildProperty>> expr) {
	var inlineValidator = new InlineValidator<TProperty>();
	return inlineValidator.RuleFor(expr);
}

```

The problem with this approach though is that once you're at the property level, the contextual informaton about the parent has been lost. Meaning this would not be possible:

```csharp
RuleForEach(s => s.Lines)
  .Property(s => s.DollarAmount)
      .GreaterThanOrEqualTo(0);
  .Property(s => s.SomethingElse)
      .NotNull();
        
```

...and instead you'd have to do this:

```csharp
RuleForEach(s => s.Lines)
  .Property(s => s.DollarAmount)
      .GreaterThanOrEqualTo(0);

RuleForEach(s => s.Lines)
  .Property(s => s.ProductName)
      .NotNull();

```

...which is more expensive as it requires iterating the collection twice, so it's not really that practical.

You could just use an inline validator directly if you don't want a separate child validator class:

```
RuleForEach(x => x.Lines).SetValidator(new InlineValidator<Order> {
  orderValidator => orderValidator.RuleFor(x => x.DollarAmount).GreaterThanOrEqualTo(0),
  orderValidator => orderValidator.RuleFor(x => x.ProductName).NotNull(0),
});
```

(Basically that's doing the same as `Property` is doing internally, but isn't then limited to a single property)

**Comment by jzabroski** [comment]
I'll have to do some digging; I was initially thinking I could use the same trick WithMessage is using, but in reverse.  In effect the builder is accumulating phantom information that isn't useful to the current builder, but could be useful for other builders (e.g., the next property).

Am I missing something obvious?

**Comment by JeremySkinner** [comment]
The problem is that once you’re at the level of an individual property, the builder interfaces would require knowledge of the parent type in order to go up to a previous level, or to jump to a different property. This would require an additional type parameter in the rule builder interfaces which is a breaking change both internally and for external consumers.

**Comment by jzabroski** [comment]
I was thinking internally it would mean a lot more tests; whether it is worth the benefit is certainly open to debate.
For external consumers, I agree, if they've created an extension method with an additional type parameter, it would break them.
However, in the general case, it would just be an extra set of APIs - but again, the downside is way more code to maintain, and I know you have explained to me the cost on your life of maintaining this library, and your life quality is a feature I can't underrate.

**Comment by JeremySkinner** [comment]
As this has been the recommended extensibility method for nearly 10 years, this would have a major impact across the user base. I don't see how this is worth breaking so many consumers when there are already alternatives that accomplish this.

**Comment by jzabroski** [comment]
I guess my thinking was, Resharper has come such a long way, I barely have to type any more to get stuff done.  I just type "rf" and get RuleFor handed to me, etc.  The whole thing is a dream world.  And I'm only on Resharper ultimate 2018.3.4 - i'm not even on 2019.1 yet. 

I guess if there was a `.SetValidator<T> where T : new, IValidator` extension method, that would let me use Resharper Create From Usage automations and be about as productive. THoughts?

**Comment by JeremySkinner** [comment]
Syntactically that'd be messy - you wouldn't be able to use type inference so all the generics would have to be specified explicitly in the call to SetValidator. 

```csharp
class FooValidator : AbstractValidator<Person> {
  public FooValidator() {
    RuleFor(x => x.Address).SetValidator<Person, string, AddressValidator>();
  }
}

public static class Ext {
	public static IRuleBuilderOptions<T, TProperty> SetValidator<T, TProperty, TValidator>(this IRuleBuilder<T,TProperty> rule) where TValidator : IValidator<TProperty>, new() {
		return rule.SetValidator(new TValidator());
	}
}
```

You're welcome to implement that in your own project if you like, but that's not something I'm interested in having in the core project. Using `SetValidator(new AddressValidator())` (or injecting it via the constructor) is much clearer,  and let's be honest, it's only a few keystrokes.

**Comment by jzabroski** [comment]
Yeah, sometimes I shake my head at how hard I try to avoid a few keystrokes.  Can you blame me?  I just finished typing up some EFCore code-first migrations for a database-first project.  Legitimately, had a tweet-worth conversation with a co-worker.

Me: Ah, too much typing.
Him: Do your hands ache?
Me: Yeah. Does that ever happen to you? (He's 10 years younger.)
Him: Yeah, only when I work with EFCore.
😆

**Comment by JeremySkinner** [comment]
Rather than chaining multiple calls to `Property` (which we've established is a massive change), an alternative would be to allow child rule definitions within a callback. This would not be a breaking change, it wouldn't require a rewrite and is also consistent with the existing syntax for both `DependentRules` and `ForEach`. It'd look something like this:

```
RuleForEach(x => x.Lines).ChildRules(line => {
  line.RuleFor(x => x.DollarAmount).GreaterThanOrEqualTo(0);
  line.RuleFor(x => x.ProductName).NotNull();
});
```

Not quite sure of the method name yet, but I think `ChildRules` makes the most sense and is conistent with `DependentRles`. Basically this would just act as a wrapper around the InlineValidator syntax I mentioned above.

Would that work for you?

**Comment by jzabroski** [comment]
That's very elegant. Yes.  You amaze me, Jeremy!

**Comment by JeremySkinner** [comment]
I'll add it for the next release. If you want to start using it before then, you can add it as an extension method:
```csharp
public static IRuleBuilderOptions<T, TProperty> ChildRules<T, TProperty>(this IRuleBuilder<T, TProperty> ruleBuilder, Action<InlineValidator<TProperty>> action) {
	if (action == null) throw new ArgumentNullException(nameof(action));
	var validator = new InlineValidator<TProperty>();
	action(validator);
	return ruleBuilder.SetValidator(validator);
}
```

**Comment by jzabroski** [comment]
Neat! Could also have a overload which "lifts" InlineValidator<TProperty> into a generic argument similar to what I asked for.  Pseudo-code would be something like:

```csharp
public static IRuleBuilderOptions<T, TProperty> ChildRules<TValidator, T, TProperty>(this IRuleBuilder<T, TProperty> ruleBuilder) where TValidator : new, InlineValidator<T>
{
	var validator = new TValidator<TProperty>();
        // magic goes here
	return ruleBuilder.SetValidator(validator);
}
```

I might play with this some to get it to work.

**Comment by JeremySkinner** [comment]
There’s not really any point. You’d lose the type inference so all the generics would have to be specified explicitly. At that point you may as well just instantiate it directly and pass it to SetValidator.

**Comment by jzabroski** [comment]
Yeah, but I have further ideas using another library to scrap that boilerplate.

**Comment by jzabroski** [comment]
Hmm.  @JeremySkinner - Any idea why this doesn't work?  I'm not sure if it's a shortcoming in the TestHelper not working on rule collections, or if it's an issue with the ChildRules extension.

Written as a Linqpad query:

```csharp
using Xunit;
using FluentValidation;
using FluentValidation.TestHelper;

void Main()
{
	new DepartmentValidatorTests().TestChildRulesValidations();	
}

public static class FluentValidationExtensions
{
	public static IRuleBuilderOptions<T, TProperty> ChildRules<T, TProperty>(this IRuleBuilder<T, TProperty> ruleBuilder, Action<InlineValidator<TProperty>> action)
	{
		if (action == null) throw new ArgumentNullException(nameof(action));
		var validator = new InlineValidator<TProperty>();
		action(validator);
		return ruleBuilder.SetValidator(validator);
	}
}
public class Person
{
	public string Name { get; set; }
}
public class Department
{
	public ICollection<Person> Persons { get; set; }
}
public class DepartmentValidator : AbstractValidator<Department>
{
	public DepartmentValidator()
	{
		RuleForEach(x => x.Persons).ChildRules(s => s.RuleFor(x => x.Name).MaximumLength(50));
	}
}
public class DepartmentValidatorTests
{
	[Fact]
	public void TestChildRulesValidations()
	{
		var badName = string.Join(string.Empty, Enumerable.Repeat("A", 51));
		var validator = new DepartmentValidator();
		validator.ShouldHaveValidationErrorFor(x => x.Persons, new List<Person>()
			{
				new Person() { Name = badName }
			});
	}
}
```

**Comment by jzabroski** [comment]
I think the issue is likely in TestHelper; to reduce complexity I changed the `DepartmentValidator` to use existing API features:

```csharp
public class DepartmentValidator : AbstractValidator<Department>
{
	public DepartmentValidator()
	{
		RuleForEach(x => x.Persons).SetValidator(new InlineValidator<Person> {
			s => s.RuleFor(x => x.Name).MaximumLength(50)
		});
	}
}
```

**Comment by jzabroski** [comment]
By the way, it works fine in the scenario when Persons is a `List<string>`.  However, when it has to go down one more level, it seems to not work any more.  I can't quite see the problem with my naked eyes at the moment.

```
void Main()
{
	new DepartmentValidatorTests().TestChildRulesValidations();	
}
public class DepartmentFlat
{
	public ICollection<string> Persons {get; set;}
}
public class DepartmentFlatValidator : AbstractValidator<DepartmentFlat>
{
	public DepartmentFlatValidator()
	{
		RuleForEach(x => x.Persons).SetValidator(new InlineValidator<string> {
			s => s.RuleFor(x => x).MaximumLength(50)
		});
	}
}
public class DepartmentValidatorTests
{
	[Fact]
	public void TestChildRulesValidations()
	{
		var badName = string.Join(string.Empty, Enumerable.Repeat("A", 51));
		var validator = new DepartmentFlatValidator();
		
		validator.ShouldHaveValidationErrorFor(x => x.Persons, new List<string>()
			{
				badName
			});
	}
}
```

**Comment by jzabroski** [comment]
NOTE: I'll probably hack a bit more on this this week, so I don't want you to stress out thinking I'm depending on you for an answer.  I suspect the issue is this line in the TestHelper logic:

```csharp
    public static IEnumerable<ValidationFailure> ShouldHaveValidationErrorFor<T, TValue>(
      this IValidator<T> validator,
      Expression<Func<T, TValue>> expression,
      TValue value,
      string ruleSet = null)
      where T : class, new()
    {
      T instanceToValidate = new T(); // T here is DepartmentFlat
      return validator.TestValidate<T, TValue>(expression, instanceToValidate, value, ruleSet, true).ShouldHaveError<T, TValue>();
    }
```

Note to self: The problem might be a bit more obvious if I switch to Michal Jankowskii's third party test helper library.

**Comment by jzabroski** [comment]
@JeremySkinner Have a look at my demoes above - I think the `InlineValidator<T>` abstraction isn't compatible with the current `FluentValidation.TestHelper` API.

**Comment by JeremySkinner** [comment]
The test helper is actually working correctly here. The rules you have defined are child rules - they're associated with the individual properties (eg "Name"), they're not associated with the whole collection, but your test is expecting the errors to be associated with the whole collection property.

**Comment by JeremySkinner** [comment]
Implemented for 8.5.

**Comment by JeremySkinner** [comment]
8.5-preview2 is now released with this change.
